feat(rust): recover remote forwarder

twittner opened the pull request:
Currently any interruption to the underlying TCP connection of a remote forwarder will permanently make it unusable. To remedy the sitution a session "medic" is added to the node manager that maintains sessions which in turn provide actions to replace themselves. Sessions are each pinged at regular frequency and upon three subsequent failures the replacement action is triggered.

Currently only remote forwarders to API nodes, i.e. with addresses like

  /dnsaddr/hostname/tcp/12345/service/api
  /ip4/1.2.3.4/tcp/5678/service/api
  /ip6/::1/tcp/12345/service/api

are considered and only if an authorised identity is present. Other addresses or adresses like these but without authorised identities will not enable recovery and not even attenpt to open a secure channel.

To allow the node manager to re-establish a secure channel the command layer no longer performs the establishment of the secure channel, but instead only project and node references are resolved and the resulting address is passed as is to the node manager. NB that this assumes that the project route does not change which is not always guaranteed. This requires making project resolution available in the manager (tbd).

twittner committed:
feat(rust): recover remote forwarder

Currently any interruption to the underlying TCP connection of a remote
forwarder will permanently make it unusable. To remedy the sitution a
session "medic" is added to the node manager that maintains sessions
which in turn provide actions to replace themselves. Sessions are each
pinged at regular frequency and upon three subsequent failures the
replacement action is triggered.

Currently only remote forwarders to API nodes, i.e. with addresses like

  /dnsaddr/hostname/tcp/12345/service/api
  /ip4/1.2.3.4/tcp/5678/service/api
  /ip6/::1/tcp/12345/service/api

are considered and only if an authorised identity is present. Other
addresses or adresses like these but without authorised identities will
not enable recovery and not even attenpt to open a secure channel.

To allow the node manager to re-establish a secure channel the command
layer no longer performs the establishment of the secure channel, but
instead only project and node references are resolved and the resulting
address is passed as is to the node manager. NB that this assumes that
the project route does not change which is not always guaranteed. This
requires making project resolution available in the manager (tbd).

SanjoDeundiak commented on implementations/rust/ockam/ockam/src/forwarder.rs:
```diff
@@ -70,6 +70,7 @@ impl Forwarder {
             forward_route,
             payload: Some(registration_payload.clone()),
         };
+        let _ = ctx.stop_worker(address.clone()).await;
```
Shouldn't we at least add prefix to the provided address, as in elixir? Not doing that looks bad even for demo purposes...
@mrinalwadhwa 

P.S. Oh, is this line the reason why forwarders to rust nodes are created without heartbeats?

twittner replied:
Sorry, this was a leftover from a previous revision. While it does not affect naming it is not relevant for this PR.

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/util/mod.rs:
```diff
@@ -244,9 +244,18 @@ impl<'a> Rpc<'a> {
-                let err = match dec.decode::<String>() {
-                    Ok(msg) => format!("Message: {msg}"),
-                    Err(_) => String::default(),
+                let err = if matches!(dec.datatype(), Ok(Type::String)) {
```
That would be great in form of a separate commit

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs:
```diff
@@ -1,37 +1,57 @@
+    /// Forwarding service is at rust node.
     #[n(3)] pub(crate) at_rust_node: bool,
+    /// Authorised identities per project.
+    #[n(4)] pub(crate) identities: HashMap<MultiAddr, IdentityIdentifier>,
```
Is empty HashMap considered no authorized identities, or any identity is authorized?

twittner replied:
The way it is currently used by the node manager is that without an authorised identity, no secure channel will be established.

SanjoDeundiak replied:
Could you please add that comment to the code?

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs:
```diff
@@ -1,37 +1,57 @@
 
 impl<'a> CreateForwarder<'a> {
-    pub fn new(address: &MultiAddr, alias: Option<String>, at_rust_node: bool) -> Self {
+    pub fn new(address: MultiAddr, alias: Option<String>, at_rust_node: bool) -> Self {
```
I would extend constructor to take all fields as arguments, that will force users to make decisions about authorized identities and credentials mode instead of default values (which are likely to cause security vulnerabilities)

SanjoDeundiak replied:
P.S. Looking at the single usage of that constructor I don't see anything against having all fields in the constructor

twittner replied:
These are still optional as creating a forwarder at another rust node for example does not require a secure channel.

SanjoDeundiak replied:
Not using Secure Channel or authorized identifiers explicitly may be fine for the time being, I'm worried about the fact that this happens implicitly. User of this code can easily create the structure, use insecure flow and not even know that those fields exist

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service.rs:
```diff
@@ -84,6 +92,7 @@ pub(crate) struct AuthorityInfo {
 
 /// Node manager provides a messaging API to interact with the current node
 pub struct NodeManager {
+    address: Address,
```
AFAIK, you can take needed address from `Context` instance, no need to store it in the field

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service.rs:
```diff
@@ -99,6 +108,14 @@ pub struct NodeManager {
+
+impl Drop for NodeManager {
+    fn drop(&mut self) {
+        self.medic.abort()
```
Should it be in shutdown function instead?

twittner replied:
I will add a shutdown method and stop the task there.

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/error.rs:
```diff
@@ -62,3 +62,9 @@ impl From<std::io::Error> for Error {
     }
 }
+
+impl From<ockam_multiaddr::Error> for Error {
```
Could be a separate commit

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/forwarder/create.rs:
```diff
@@ -46,48 +51,74 @@ impl CreateCommand {
-        rpc.request(req(&cmd, at_rust_node)?).await?;
-        rpc.parse_and_print_response::<ForwarderInfo>()?;
-        Ok(())
+async fn rpc(ctx: Context, (opts, cmd): (CommandGlobalOpts, CreateCommand)) -> Result<()> {
```
I suspect this code can be used for other commands as well, would be great to move it to some shared place in the future

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/forwarder/create.rs:
```diff
@@ -46,48 +51,74 @@ impl CreateCommand {
-    } else {
-        cmd.forwarder_name.clone()
+    let req = {
+        let alias = if at_rust_node {
```
Would be great to unify that, this is out of context of this PR though

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/forwarder/create.rs:
```diff
@@ -46,48 +51,74 @@ impl CreateCommand {
+            cmd.forwarder_name.clone()
+        };
+        let mut body = CreateForwarder::new(ma, Some(alias), at_rust_node);
+        body.set_identities(pa)
```
Should we allow empty HashMap here?

twittner committed:
fix(rust): cleanup

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session/sessions.rs:
```diff
@@ -0,0 +1,175 @@
+#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash, Encode, Decode)]
+#[rustfmt::skip]
+pub struct Key {
+    #[n(0)] ctr: u64,
```
What is the point of the counter?

twittner replied:
The key is used to index sessions. The counter gives a strongly monotonic number and avoids collisions which a purely random key could cause.

SanjoDeundiak replied:
Isn't that potentially leaking data about number of sessions of this node to the other node?

twittner replied:
Not potentially, but it definitely exposes the number of sessions. If that is a concern we would have to resort to just random keys and perform a collision check when generating it. If preferred I can implement that change.

SanjoDeundiak replied:
Yeah, I think leaking such a counter is a bad idea. How about random keys that are long enough so you don't worry about collisions?

SanjoDeundiak replied:
BTW we can have `ctr` and `rnd` if you like that, it's just usage should be vice versa: `rnd` is unique identifier of the session, while `ctr` is an intra-session ping counter

twittner replied:
I switched to a 24-bytes random key.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+
+impl ockam_core::Message for Message {}
+
+/// A collector receives messages from a [`Responder`] and forwards them.
```
What is Responder here?

twittner replied:
Another leftover from before https://github.com/build-trust/ockam/pull/3485.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+                        let sender = ctx.clone();
+                        self.pings
+                            .spawn(async move { (key, sender.forward(l).await) });
+                        continue;
```
`else if` is cleaner IMHO

adrianbenavides commented on implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs:
```diff
@@ -1,44 +1,89 @@
+                                    auth,
+                                    req.mode,
+                                );
+                                self.sessions.lock().unwrap().add(s);
```
I've seen that you always unwrap the `lock()` result. Is this done on purpose to let the mutex in a "poisoned" state if something panics?

twittner replied:
It is sort of the default thing to do unless one is sure that no invariants are broken if a panic occurs while holding the lock. While panics are not expected when dealing with sessions they are propagated across thread boundaries this way should they arise.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+        self.go(ctx, rx).await
+    }
+
+    async fn go(mut self, ctx: Context, mut rx: mpsc::Receiver<Message>) -> ! {
```
This function would benefit from some comments

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+    async fn get_results(&mut self, rx: &mut mpsc::Receiver<Message>) {
+        loop {
+            tokio::select! {
+                p = self.pings.join_next(), if !self.pings.is_empty() => match p {
```
Why `if !self.pings.is_empty()` check is needed?

twittner replied:
`JoinSet::join_next` immediately returns `None` if empty. By checking first if the set is empty, the future returned from `JoinSet::join_next` is not evaluated, leaving the other branches. This makes the whole loop less busy.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+
+    async fn get_results(&mut self, rx: &mut mpsc::Receiver<Message>) {
+        loop {
+            tokio::select! {
```
Could you please elaborate on how this works? Is that intended that we try to run pings and session reestablishments and receive pongs concurrently? Shouldn't we do that in parallel? Also, `tokio::select` is meant to cancel all futures except the one that finished. So, in a loop, we start polling all of the futures I mentioned above one-by-one (allowing them to make some progress), wait till one of them finishes, abort everything else (and remove all the progress that they may achieved, also not cleaning up any state they may have changed so far), then start all of them again on next loop iteration. Am I wrong about how this works?

twittner replied:
Operationally there are ping tasks and replacement tasks that execute in the context of tokio `JoinSet`s. Their progress is driven by the tokio executor. The `tokio::select!` call only evaluates cancellation-safe futures which eventually yield a result from those tasks (or a message in case of `mpsc::Receiver::recv`). No progress is removed and no state needs to be cleaned up when the loosing futures care cancelled.

For details see:
- https://docs.rs/tokio/latest/tokio/macro.select.html
- https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html

SanjoDeundiak replied:
@twittner oh, I guess `JoinSet` is the owner of the futures in this case, so futures are not cancelled in `tokio::select`. Tnx, for the clarification. Do you know if futures are able to progress in the background, or only while `join_next` is called? That wasn't clear for me after reading the doc

SanjoDeundiak replied:
Also, what are you thoughts about having centralized Medic that tracks all sessions (as in this PR) vs having dedicated Medic per session? Is there a chance latter approach would look simpler?

twittner replied:
> Do you know if futures are able to progress in the background, or only while `join_next` is called?

A `JoinSet` spawns the futures as regular tokio tasks and they make progress independently of the `join_next` call.

> Also, what are you thoughts about having centralised Medic that tracks all sessions (as in this PR) vs having dedicated Medic per session?

I think tracking all sessions here is more flexible, for instance when expressing dependencies. Isolated session tracking in actors makes such reasoning comparatively difficult. For example, when a secure channel is shared by multiple sessions, it would be enough to ping it once and restart all sessions depending on it. This is easier to implement when all sessions are treated as data in a single `Medic`.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/session.rs:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    async fn get_results(&mut self, rx: &mut mpsc::Receiver<Message>) {
```
Require some refactoring in the future, it's a little bit hard to read

SanjoDeundiak reviewed (changes requested):
Still reviewing, submitting what I have so far...

twittner committed:
fix(rust): changes due to review comments

- Abort medic task in `NodeManager::shutdown`.
- Use address from `Context` instead of storing it as a member variable.
- Comment `Medic::go` method.
- Use if/else instead of continue in `Medic::go`.
- Fix `Collector` comment.
- Made `forwarder::connect` a method of `NodeManager`.
- Commented on authorised identities.
- Made session `Key` entirely random.
- Remote `CreateForwarder::{set_identities,set_credentials_mode}`.
- Use `multiaddr_to_route` for secure channel address.

Also lowered the log level from debug to trace for several statements to
not spam the node log.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service.rs:
```diff
@@ -237,6 +250,11 @@ impl NodeManager {
             registry: Default::default(),
+            medic: {
+                let ctx = ctx.async_try_clone().await?;
+                tokio::spawn(medic.start(ctx))
```
We ideally want to limit direct usage of tokio to `ockam_node` crate. It's totally fine for now though, just FYI

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs:
```diff
@@ -49,14 +94,135 @@ impl NodeManager {
+    authorised: &HashMap<MultiAddr, IdentityIdentifier>,
+    mode: CredentialExchangeMode,
+) -> Result<Address> {
+    debug!(%addr, %prev, "recreating secure channel");
```
Is sending a message to itself is better than simple function call?

twittner replied:
Most methods of `NodeManager` require unique access (`&mut self`). The recovery futures run asynchronously and we would have to wrap the manager into an `Arc<Mutex<_>>` to apply those methods. Such a change would be much larger so I chose the message sending instead.

SanjoDeundiak replied:
Yeah, probably there should be `NodeManager` that is doing all the heavy lifting, and `NodeManagerWorker` that is a Worker that delegates calls to `NodeManager`. In this case either `NodeManager` can be wrapped inside `Arc<Mutex>` of its fields could be turned into thread-safe structures. Definitely out of the context of this PR though. Thoughts?

twittner replied:
That would be a good change I think.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs:
```diff
@@ -49,14 +94,135 @@ impl NodeManager {
+    }
+    let res: CreateSecureChannelResponse = d.decode()?;
+    let mad = res.addr()?;
+    if let Some(p) = mad.first() {
```
Isn't in cleaner to convert this multiaddr to a route (we have a function for that) and check that there is only one hop?

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs:
```diff
@@ -49,14 +94,135 @@ impl NodeManager {
+}
+
+/// Create a secure channel to an API node or return the address route as is.
+async fn connect(
```
Should it be a NodeManager's function?

twittner replied:
I deliberately made it a private function of this module for the time being so that other parts of the `NodeManager` would not be tempted to use it.

SanjoDeundiak replied:
But it can be private even if it's a NodeManager's function

twittner replied:
Not sure I get what you mean. If it is a private *method* of the `NodeManager` it would be available in the whole scope of it vs. just being a private function of this module. Did you have something else in mind?

SanjoDeundiak replied:
AFAIK private function of the struct, that is defined in submodule, is only available from that submodule. In other words this `connect` function would be available from this file only doesn't matter if it's a static function of a struct's function

twittner replied:
Thanks, I did not think it would behave like this but it makes sense.

twittner commented:
@SanjoDeundiak: Anything left that needs to be addressed?

SanjoDeundiak commented:
@mergifyio rebase
